src/tools/w3dviewer/mainfrm.cpp

```diff
@@ -0,0 +1,1611 @@
+
+        if (ret) {
+            CWnd *pane = m_splitter.GetPane(0, 1);
+
+            if (pane != nullptr) {
+                char path[MAX_PATH];
+                GetModuleFileName(nullptr, path, MAX_PATH);
+                char *str = strrchr(path, '\\');
+
+                if (str != nullptr) {
+                    *str = '\0';
+                    SetCurrentDirectory(path);
+                }
+
+                W3D::Init(pane->m_hWnd);
+                W3D::Enable_Static_Sort_Lists(true);
+                DeviceWidth = theApp.GetProfileInt("Config", "DeviceWidth", 640);
+                DeviceHeight = theApp.GetProfileInt("Config", "DeviceHeight", 480);
+                GetDevice(false);
+                W3D::Enable_Munge_Sort_On_Load(AfxGetApp()->GetProfileInt("Config", "MungeSortOnLoad", 0) == 1);
+                W3D::Enable_Sorting(AfxGetApp()->GetProfileInt("Config", "EnableSorting", 1) == 1);
+            }
+        }
+    }
+
+    return ret;
+}
+
+void CMainFrame::RestoreWindowPos()
+{
+    int left = theApp.GetProfileInt("Window", "Left", -1);
+    int right = theApp.GetProfileInt("Window", "Right", -1);
+    int top = theApp.GetProfileInt("Window", "Top", -1);
+    int bottom = theApp.GetProfileInt("Window", "Bottom", -1);
+    bool maximized = theApp.GetProfileInt("Window", "Maximized", -1) == 1;
+
+    if (left != -1 && right != -1 && top != -1 && bottom != -1) {
+        if (maximized) {
+            ShowWindow(SW_SHOWMAXIMIZED);
+        } else {
+            SetWindowPos(nullptr, left, top, right - left, bottom - top, SWP_NOZORDER);
+        }
+    }
+}
+
+void CMainFrame::DoProperties()
+{
+    // do later
+}
+
+void CMainFrame::UpdateMenus(AssetType type)
+{
+    if (m_currentType != type) {
+        if (m_currentType >= ASSET_TYPE_HIERARCHY
+            && (m_currentType <= ASSET_TYPE_ANIMATION || m_currentType == ASSET_TYPE_MORPH)) {
+            CMenu *menu = GetMenu();
+
+            if (menu != nullptr) {
+                menu->RemoveMenu(6, MF_BYPOSITION | MF_DELETE);
+                DrawMenuBar();
+            }
+
+            if (m_currentType == ASSET_TYPE_ANIMATION || m_currentType == ASSET_TYPE_MORPH) {
+                m_animationToolbarVisible = m_animationToolbar.IsWindowVisible();
+                ShowControlBar(&m_animationToolbar, FALSE, FALSE);
+            }
+        }
+
+        switch (type) {
+            case ASSET_TYPE_HIERARCHY: {
+                CMenu *menu = GetMenu();
+
+                if (menu != nullptr) {
+                    MENUITEMINFO info;
+                    memset(&info, 0, sizeof(info));
+                    info.cbSize = sizeof(info);
+                    info.hSubMenu = GetSubMenu(LoadMenu(AfxGetResourceHandle(), MAKEINTRESOURCE(IDR_HIERARCHYMENU)), 0);
+                    info.fMask = MIIM_TYPE | MIIM_DATA | MIIM_SUBMENU;
+                    info.fType = MF_STRING;
+                    info.dwTypeData = "&Hierarchy";
```

Comment:
Ok so I looked at the MFC source code to see how it did this and it does an explicit cost_cast<LPSTR> to remove the constness. This then compiles fine with /Zc:strictStrings enabled (no warning) and since its in the official MFC source code from Microsoft is clearly an acceptable way to do it.